core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,112 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ *
+ * Note that although the preimage and image contain only moved points, the
+ * domain and range of a Perm is all non-negative integers.
+ */
+class Perm private(private val mapping: Map[Int, Int]) extends (Int => Int) {
+  require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
```

Comment:
Would it make sense to move checks into smart constructors since actual constructor is private?

Reply:
Ohhh, yes. That's a good idea. Then we could optimize all the internal methods to skip checks.